embedded/remotestorage/s3/s3.go

```diff
@@ -691,4 +715,126 @@ func (s *Storage) scanObjectNames(ctx context.Context, prefix string, limit int)
 	return entries, subPaths, nil
 }
 
+func (s *Storage) getRoleCredentials() error {
+	if !s.S3RoleEnabled {
+		return nil
+	}
+
+	if s.accessKeyID != "" || s.secretKey != "" {
+		return ErrKeyCredentialsProvided
+	}
+
+	var err error
+	s.accessKeyID, s.secretKey, err = s.requestCredentials()
+	if err != nil {
+		return err
+	}
+
+	s3CredentialsRefreshTicker := time.NewTicker(s.awsCredsRefreshPeriod)
+	go func() {
+		select {
+		case _ = <-s3CredentialsRefreshTicker.C:
+			accessKeyID, secretKey, err := s.requestCredentials()
+			if err != nil {
+				log.Printf("S3 role credentials lookup failed with an error: %v", err)
+			} else {
+				s.accessKeyID, s.secretKey = accessKeyID, secretKey
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (s *Storage) requestCredentials() (string, string, error) {
+	tokenReq, err := http.NewRequest("PUT", fmt.Sprintf("%s%s",
+		s.awsInstanceMetadataURL,
+		"/latest/api/token",
+	), nil)
+	if err != nil {
+		return "", "", errors.New("cannot form metadata token request")
+	}
+
+	tokenReq.Header.Set("X-aws-ec2-metadata-token-ttl-seconds", "21600")
```

Comment:
is this header also useful in most S3 storages? I mean, when not using aws

Reply:
same questions for `/latest/api/token` and `/latest/meta-data/iam/info`

Reply:
This is specific to AWS and S3, yes, but this is the flow that is specifically applicable to AWS.
I checked Minio docs and references, I didn't find any way to use role-based auth. Essentially, it means that role-based auth and the whole `S3RoleEnabled` is AWS specific.

Should I refactor this? I could spend some time and make it _in theory_ portable while the reality is that there are no other applications for this code but AWS S3.